File: aat-lib/src/main/java/ch/bailu/aat_lib/service/cache/Obj.java
package ch.bailu.aat_lib.service.cache;

import javax.annotation.Nonnull;

import ch.bailu.aat_lib.app.AppContext;
import ch.bailu.aat_lib.logger.AppLog;
import ch.bailu.foc.Foc;
import ch.bailu.foc.FocName;


public abstract class Obj implements ObjBroadcastReceiver {

    public static final int MIN_SIZE=100;

    private final String ID;

    private long accessTime=System.currentTimeMillis();
    private int lock=0;

    private Exception exception = null;

    public Obj(String id) {
        ID=id;
    }


    @Nonnull
    @Override
    public String toString() {
        return ID;
    }


    public String getID() {
        return ID;
    }

    public Foc getFile() {
        AppLog.w(this, "Default implementation of getFile() called!");
        return new FocName(ID);
    }


    protected void setException(Exception e) {
        exception = e;
    }


    public boolean hasException() {
        return exception != null;
    }


    public Exception getException() {
        return exception;
    }


    public boolean isLocked() {
        return lock > 0;
    }


    public void onInsert(AppContext appContext) {}
    public void onRemove(AppContext appContext) {}


    public synchronized void lock(AppContext appContext) {
        lock++;
        access();
    }

    public synchronized void free() {
        lock--;
    }


    public boolean isReadyAndLoaded() {
        return true;
    }


    public boolean isLoaded() {
        return isReadyAndLoaded() || hasException();
    }


    public abstract long getSize();

    public synchronized void access() {
        accessTime=System.currentTimeMillis();
    }

    public synchronized long getAccessTime() {
        return accessTime;
    }


    @Override
    public int hashCode() {
        return toString().hashCode();
    }

    public static class Factory {
        public Obj factory(String id, AppContext appContext) {
            return ObjNull.NULL;
        }
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/service/cache/Obj.kt
package ch.bailu.aat_lib.service.cache

import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.logger.AppLog.w
import ch.bailu.foc.Foc
import ch.bailu.foc.FocName
import javax.annotation.Nonnull

abstract class Obj(private val id: String) : ObjBroadcastReceiver {
    private var accessTime = System.currentTimeMillis()
    private var lock = 0
    private var exception: Exception? = null

    @Nonnull
    override fun toString(): String {
        return id
    }

    fun getID(): String {
        return id
    }

    open fun getFile(): Foc {
        w(this, "Default implementation of getFile() called!")
        return FocName(id)
    }

    protected fun setException(e: Exception?) {
        exception = e
    }

    fun hasException(): Boolean {
        return exception != null
    }

    fun getException(): Exception? {
        return exception
    }

    fun isLocked(): Boolean {
        return lock > 0
    }

    open fun onInsert(appContext: AppContext) {}
    open fun onRemove(appContext: AppContext) {}

    @Synchronized
    fun lock() {
        lock++
        access()
    }

    @Synchronized
    fun free() {
        lock--
    }

    open fun isReadyAndLoaded(): Boolean {
        return true
    }

    open fun isLoaded(): Boolean {
        return isReadyAndLoaded() || hasException()
    }

    abstract fun getSize(): Long

    @Synchronized
    open fun access() {
        accessTime = System.currentTimeMillis()
    }

    @Synchronized
    fun getAccessTime(): Long {
        return accessTime
    }

    override fun hashCode(): Int {
        return toString().hashCode()
    }

    open class Factory {
        open fun factory(id: String, appContext: AppContext): Obj {
            return ObjNull
        }
    }

    companion object {
        const val MIN_SIZE: Int = 100
    }
}
